pywr/parameters/_parameters.pyx

```diff
@@ -1906,6 +1906,55 @@ cdef class PiecewiseIntegralParameter(Parameter):
 PiecewiseIntegralParameter.register()
 
 
+cdef class FlowDelayParameter(Parameter):
+    """Parameter that returns the delayed flow for a node after a given number of timesteps or days
+    
+    Parameters
+    ----------
+    model : `pywr.model.Model`
+    node: Node
+        The node to delay for.
+    timesteps: int
+        Number of timesteps to delay the flow.
+    days: int
+        Number of days to delay the flow. Specifying a number of days (instead of a number
+        of timesteps) is only valid if the number of days is exactly divisible by the model timestep length.
+    """
+
+    def __init__(self, model, node, *args, **kwargs):  
+        self.node = node
+        self.timesteps = kwargs.pop('timesteps', 0)
+        self.days = kwargs.pop('days', 0)
+        super().__init__(model, *args, **kwargs)
+
+    cpdef setup(self):
+        super(FlowDelayParameter, self).setup()
+        cdef int r
+        if self.days > 0:
+            r = self.days % self.model.timestepper.delta
+            if r == 0:
+                self.timesteps = self.days / self.model.timestepper.delta
+            else:
+                raise ValueError('The delay defined as number of days is not exactly divisible by the timestep delta.')
+        if self.timesteps < 1:
+            raise ValueError('The number of time-steps for a FlowDelayParameter node must be greater than one.')
+        self._memory = np.zeros((self.timesteps,  len(self.model.scenarios.combinations)))
+
+    cpdef double value(self, Timestep ts, ScenarioIndex scenario_index) except? -1:
+        return self._memory[-1, scenario_index.global_id]
+
+    cpdef after(self):
+        self._memory = np.roll(self._memory, 1, axis=0)
```

Comment:
I think an internal pointer would be more efficient, but happy to be proved wrong. This is probably more readable.

Reply:
Some crude tests show that `np.roll` is about 10% slower than using a pointer so I've switched it to use a pointer